integration/image/delta_test.go

```diff
@@ -143,3 +148,402 @@ func pullBaseAndTargetImages(t *testing.T, client apiclient.APIClient, base, tar
+	for _, r := range allSizes {
+		change := ""
+		if r.got < r.want {
+			change = "IMPROVED!"
+		} else if r.got > r.want {
+			change = "BAD!"
+		}
+		t.Logf(format, r.desc, r.want, r.got, change)
+	}
+	t.Log("-------------------------------------------------------------")
+}
+
+// TestDeltaCorrectness checks if applying a delta on a base image results in an
+// image with the same contents as the original target image.
+func TestDeltaCorrectness(t *testing.T) {
+	defer setupTemporaryTestRegistry(t)()
+
+	client := testEnv.APIClient()
+	ctx := context.Background()
+
+	for _, tc := range deltaTestCases {
+		delta := deltaName(tc.base, tc.target)
+		t.Run(delta, func(t *testing.T) {
+			imagesToBuild := []string{tc.base, tc.target}
+			if tc.images != nil {
+				imagesToBuild = tc.images
+			}
+
+			imagesToRemove := []string{tc.target}
+			if tc.removeBeforePull != nil {
+				imagesToRemove = tc.removeBeforePull
+			}
+
+			// Build all required images
+			var imageRemovers []func()
+			for _, image := range imagesToBuild {
+				removeImage := ttrBuildImage(ctx, t, client, image)
+				if sliceContains(imagesToRemove, image) {
+					imageRemovers = append(imageRemovers, removeImage)
+				} else {
+					defer removeImage()
+				}
+			}
+
+			// Create delta of them and push this delta.
+			removeDelta := ttrCreateDelta(ctx, t, client, tc.base, tc.target)
+			ttrPushImage(ctx, t, client, delta)
+
+			// The delta we have locally shall not be the same as the target image.
+			targetHash := ttrImageHash(ctx, t, client, tc.target)
+			deltaHash := ttrImageHash(ctx, t, client, delta)
+			assert.Assert(t, !reflect.DeepEqual(targetHash, deltaHash))
+
+			// Remove the delta and the target image.
+			for _, removeImage := range imageRemovers {
+				removeImage()
+			}
+			removeDelta()
+
+			// Pull the delta. This will cause it to be applied.
+			ttrPullImage(ctx, t, client, delta)
+
+			// The (now applied) delta shall be the same the target image was.
+			appliedDeltaHash := ttrImageHash(ctx, t, client, delta)
+			assert.Assert(t, reflect.DeepEqual(targetHash, appliedDeltaHash))
+		})
+	}
+}
+
+//
+// Temporary Test Registry (TTR) helper functions
+//
+// With regards to image names: all these functions expect plain image names
+// (like "002"), but under the hood they'll add the temporary test registry name
+// to the images ("127.0.0.1:5000/002").
+//
+// In other words, this is about image tagging. Most of these functions will
+// work even without the temporary test registry running. We just need the
+// registry for push and pull operations.
+//
+
+// ttrBuildImage builds a given image and tags it. Returns a function that, when
+// called, remove the image built.
+func ttrBuildImage(ctx context.Context, t *testing.T, client apiclient.APIClient, image string) func() {
+	source := fakecontext.New(t, "")
+	defer source.Close()
+
+	copy.DirCopy("../testdata/delta/", source.Dir, copy.Content, true)
```

Comment:
this fails for me with `operation not supported`, I ran:
```
$ make test-integration TEST_SKIP_INTEGRATION_CLI=1 TEST_FILTER=DeltaSize
```

Reply:
When run with your command line I get different (but possibly related) errors:

```
    --- FAIL: TestDeltaSize/delta-000-001 (0.05s)
        delta_test.go:407: assertion failed: error is not nil: Error response from daemon: Cannot locate specified Dockerfile: 000.Dockerfile
    --- FAIL: TestDeltaSize/delta-000-002 (0.03s)
        delta_test.go:407: assertion failed: error is not nil: Error response from daemon: Cannot locate specified Dockerfile: 000.Dockerfile
... etc ...
```

I was also passing `BIND_DIR=.` to run, but I think requiring it shouldn't be necessary. I'll take a look.

Reply:
Hmm yeah that's the error I got at first, then I added a check for the error that `DirCopy` returns and got `operation not supported`, which I was able to resolve by passing `false` with the third parameter (`copyXattrs`)